inception-concept-linking/src/main/java/de/tudarmstadt/ukp/inception/conceptlinking/util/QueryUtil.java

```diff
@@ -67,58 +68,146 @@
+        String exactMatchingMention = getExactMatchingQueryPart("exactMention");
+
         String query = String.join("\n",
             "DEFINE input:inference 'instances'",
             SPARQL_PREFIX,
-            "SELECT DISTINCT ?e2 ?altLabel ?label ?description WHERE",
+            "SELECT DISTINCT ?e2 ?label ?description WHERE",
             "{",
             "  {",
+            exactMatchingTypedString,
+            "  } ",
+            "  UNION",
+            "  {",
+            exactMatchingMention,
+            "  }",
+            "  FILTER EXISTS { ?e2 ?p ?v }",
+            "  FILTER NOT EXISTS ",
+            "  {",
+            "    VALUES ?topic {" + String.join(" ", WIKIMEDIA_INTERNAL,
+                WIKIMEDIA_PROJECT_PAGE, WIKIMEDIA_CATEGORY, WIKIMEDIA_DISAMBIGUATION_PAGE,
+                WIKIMEDIA_LIST_ARTICLE, WIKIMEDIA_TEMPLATE, WIKIMEDIA_NEWS_ARTICLE,
+                WIKIMEDIA_NAVIGATIONAL_TEMPLATE) +
+                "}",
+            "    ?e2 rdf:type ?topic",
+            "  }",
+            "  ?e2 rdfs:label ?label.",
+            "  FILTER ( lang(?label) = \"en\" )",
+            "}");
+
+        ValueFactory vf = SimpleValueFactory.getInstance();
+
+        TupleQuery tupleQuery = conn.prepareTupleQuery(QueryLanguage.SPARQL, query);
+        tupleQuery.setBinding("exactTyped", vf.createLiteral(aTypedString));
+        tupleQuery.setBinding("exactMention", vf.createLiteral(aMention));
+
+        tupleQuery.setBinding("descriptionIri", aDescriptionIri);
+        return tupleQuery;
+    }
+
+    private static String getFullTextMatchingQueryPart(String aString, int aLimit) {
+        return  String.join("\n",
+            "    SELECT DISTINCT ?e2 ?altLabel ?description WHERE",
             "    {",
             "      VALUES ?labelpredicate {rdfs:label skos:altLabel}",
-            "      GRAPH " + TERMS,
             "      {",
             "        ?e2 ?labelpredicate ?altLabel.",
-            "        ?altLabel bif:contains '?entityLabel'. ",
+            "        ?altLabel bif:contains '?" + aString + "'. ",
             "        OPTIONAL",
             "        {",
             "          ?e2 ?descriptionIri ?description.",
             "          FILTER ( lang(?description) = \"en\" )",
             "        }",
             "      }",
             "    }",
+            "    LIMIT " + aLimit);
+    }
+
+    /**
+     *
+     * This query retrieves candidates via full-text matching of their labels and full-text-search
+     *
+     * @param aString String for which to perform full text search
+     * @param aLimit maximum number of results
+     * @param aDescriptionIri KB-specific IRI that indicates a description
+     * @return a query to retrieve candidate entities
+     */
+    public static TupleQuery generateCandidateFullTextQuery(RepositoryConnection conn,
+        String aString, int aLimit, IRI aDescriptionIri)
+    {
+        aString = RenderUtils.escape(aString).toLowerCase(Locale.ENGLISH);
```

Comment:
Create a local variable with a new name instead of assigning a value to the parameter variable. Parameters should not be reassigned.